File: Top Interview Questions/Array/33. Search in Rotated Sorted Array.java
class Solution {
    public int search(int[] nums, int target) {
        int rotated = findRotated(nums);
        
        // int rotated = 4;
        /*
        Binary search
        */
        int left = 0, right = rotated - 1;
        while (left <= right) {
            int pivot = (left + right) / 2;
            if (nums[pivot] == target)
                return pivot;
            else {
                if (target < nums[pivot])
                    right = pivot - 1;
                else
                    left = pivot + 1;
            }
        }
        
        left = rotated; right = nums.length - 1;
        while (left <= right) {
            int pivot = (left + right) / 2;
            if (nums[pivot] == target)
                return pivot;
            else {
                if (target < nums[pivot])
                    right = pivot - 1;
                else
                    left = pivot + 1;
            }
        }
        
        return -1;
    }
    
    public static int findRotated(int[] nums) {
        int left = 0, right = nums.length - 1;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            //System.out.print(left);
            //System.out.print("\t");
            //System.out.println(right);
            if (nums[left] <= nums[right]) {
                return left;
            } else {
                if (nums[left] <= nums[mid]) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
        }
        return left;
    }
    
}

File: Top Interview Questions/Array/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # find pivot
        l, r = 0, len(nums) - 1
        pivot = 0
        while l <= r:
            m = (l + r) // 2
            if nums[l] <= nums[r]:
                break
            else:
                if nums[l] <= nums[m]:
                    l = m + 1
                else:
                    r = m
        pivot = l
        
        # bi search left
        l, r = 0, pivot - 1
        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            elif nums[m] < target:
                l = m + 1
            else:
                r = m - 1
        # bi search right
        l, r = pivot, len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            elif nums[m] < target:
                l = m + 1
            else:
                r = m - 1
                
        return -1
        